gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/TaskDependsOn.java

```diff
@@ -0,0 +1,135 @@
+        need `classes`. It also doesn't encourage you to write tasks with explicit, fine-grained inputs and outputs.
+        Having unnecessary dependencies adds unnecessary work to a build and hurts task parallelism.
+
+        However, there are cases where `dependsOn` is unavoidable:
+        1. Wiring up tasks to lifecycle tasks e.g. `build.dependsOn(myTask)`
+        2. Tasks that do system-wide setup like installing npm. In these cases, we recommended that you specify the
+            installation directory as an `@OutputDirectory` of the task anyways.
+
+        This errorprone tries to detect legitimate uses of `dependsOn`. You can suppress false positives, but we
+        encourage you to do so sparingly, and think about whether something can be improved about your task design.
+        """)
+public final class TaskDependsOn extends GradleGuideBugChecker implements BugChecker.MethodInvocationTreeMatcher {
+
+    private static final Set<String> KNOWN_LIFECYCLE_TASKS = Set.of(
+            // Core bundling tasks
+            "org.gradle.api.tasks.bundling.Jar",
+            "org.gradle.api.tasks.bundling.War",
+            "org.gradle.api.tasks.bundling.Zip",
+            "org.gradle.api.tasks.bundling.Tar",
+
+            // Testing
+            "org.gradle.api.tasks.testing.Test",
+
+            // Compilation
+            "org.gradle.api.tasks.compile.JavaCompile",
+
+            // Documentation
+            "org.gradle.api.tasks.javadoc.Javadoc",
+
+            // Application
+            "org.gradle.api.tasks.JavaExec",
+            "org.gradle.api.tasks.application.CreateStartScripts",
+
+            // Quality
+            "org.gradle.api.plugins.quality.Checkstyle",
+            "com.github.spotbugs.snom.SpotBugsTask",
+
+            // Publishing
+            "org.gradle.api.publish.maven.tasks.PublishToMavenRepository",
+            "org.gradle.api.publish.maven.tasks.PublishToMavenLocal",
+            "org.gradle.api.publish.maven.tasks.GenerateMavenPom",
+
+            // Cleanup
+            "org.gradle.api.tasks.Delete");
+
+    private static final Matcher<ExpressionTree> TASK_DEPENDS_ON = MethodMatchers.instanceMethod()
+            .onDescendantOf("org.gradle.api.Task")
+            .named("dependsOn");
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        if (!TASK_DEPENDS_ON.matches(tree, state)) {
+            return Description.NO_MATCH;
+        }
+
+        Optional<ExpressionTree> receiverMaybe = Optional.ofNullable(ASTHelpers.getReceiver(tree));
+        if (receiverMaybe.isEmpty()) {
+            return Description.NO_MATCH;
+        }
+        ExpressionTree receiver = receiverMaybe.get();
+        Type taskType = ASTHelpers.getType(receiver);
+
+        // It's alright to wire up lifecycle tasks with dependsOn
+        if (isKnownLifecycleTask(taskType, state)) {
+            return Description.NO_MATCH;
+        }
```

Comment:
yeah, given the impl, this is just not true

e.g. for a lot of cases, it's weird that `jar` has a depends on, it means that something has gone wrong else where.

It kinda sounds like you're conflating tasks for implementations you don't control, and the actual meaning of lifecycle tasks.

From [Organizing tasks](https://docs.gradle.org/current/userguide/organizing_tasks.html):

>Actionable tasks in Gradle are tasks that perform actual work, such as compiling code. Lifecycle tasks are tasks that do not do work themselves. These tasks have no actions, instead, they bundle actionable tasks and serve as targets for the build.

Reply:
yep you're right, I was conflating the two. lets remove everythng that has to do with `KNOWN_LIFECYCLE_TASK` here entirely.